File: geekbang/algorithm/source-code/1/lc150 逆波兰表达式求值.js
/**
 * @param {string[]} tokens
 * @return {number}
 */
var evalRPN = function(tokens) {
    var calc = function(x, y, op) {
        if (op === "+") return x + y;
        if (op === "-") return x - y;
        if (op === "*") return x * y;
        // (x/y)>>0是整除，正负数均可使用。Math.floor只能用于正数
        if (op === "/") return (x / y) >> 0;
        return 0;
    }

    let s = [];
    for (let token of tokens) {
        if (token === "+" || token === "-" || token === "*" || token === "/") {
            let y = s[s.length - 1];
            s.pop();
            let x = s[s.length - 1];
            s.pop();
            let z = calc(x, y, token);
            s.push(z);
        } else {
            s.push(parseInt(token)); // 字符串转化为整数
        }
    }
    return s[s.length - 1];
};

File: geekbang/algorithm/source-code/1/lc150 逆波兰表达式求值.cpp
class Solution {
public:
    long long evalRPN(vector<string>& tokens) {
        for (string& token : tokens) {
            if (token == "+" || token == "-" || token == "*" || token == "/") {
                long long y = s.top();
                s.pop();
                long long x = s.top();
                s.pop();
                long long z = calc(x, y, token);
                s.push(z);
            } else {
                s.push(atoi(token.c_str())); // 字符串转化为整数
            }
        }
        return s.top();
    }

private:
    stack<long long> s;

    long long calc(long long x, long long y, string& op) {
        if (op == "+") return x + y;
        if (op == "-") return x - y;
        if (op == "*") return x * y;
        if (op == "/") return x / y;
        return 0;
    }
};